Mobile/ShopLens/Droid/Activities/MainActivity.cs

```diff
@@ -103,59 +115,145 @@ protected override void OnCreate(Bundle savedInstanceState)
+
+            noTalkBack = true;
         }
 
-        // When the current voice recognition session stops.
-        public void OnResults(Bundle results)
+        private bool IsTutorialNeeded()
         {
-            var matches = results.GetStringArrayList(SpeechRecognizer.ResultsRecognition);
-            if (matches.Count > 0)
+            bool firstTime = prefs.GetBoolean("FirstTime", true);
+
+            if (firstTime)
             {
-                if (matches[0] == ConfigurationManager.AppSettings["CmdOpenCamera"])
-                {
-                    var intent = new Intent(this, typeof(CameraActivity));
-                    StartActivity(intent);
-                }
-                else if (matches[0] == ConfigurationManager.AppSettings["CmdOpenCart"])
-                {
-                    var intent = new Intent(this, typeof(ShoppingCartActivity));
-                    StartActivity(intent);
-                }
-                else if (matches[0] == ConfigurationManager.AppSettings["CmdOpenList"])
-                {
-                    var intent = new Intent(this, typeof(ShoppingListActivity));
-                    StartActivity(intent);
-                }
-                // For debugging purposes.
-                else
-                {
-                    voiceCommandButton.Text = matches[0];
-                }
+                prefs.Edit().PutBoolean("FirstTime", false).Commit();
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        private void RunUserTutorial()
+        {
+            var message = ConfigurationManager.AppSettings["InitialTutorialMsg"];
+            tts.Speak(message, QueueMode.Flush, null, needUserAnswerId);
+        }
+
+        private void ContinueUserTutorial()
+        {
+            var message = ConfigurationManager.AppSettings["FollowUpTutorialMsg"];
+            tts.Speak(message, QueueMode.Flush, null, needUserAnswerId);
+        }
+
+        // TTS Engine method called when TTS is initialized.
+        public void OnInit([GeneratedEnum] OperationResult status)
+        {
+            if (status == OperationResult.Error)
+            {
+                tts.SetLanguage(Locale.Default);
+            }
+            else if (status == OperationResult.Success)
+            {
+                tts.SetLanguage(Locale.Us);
+            }
+
+            if (tutorialNeeded)
+            {
+                RunUserTutorial();
+            }
+            else
+            {
+                var welcomeMsg = ConfigurationManager.AppSettings["WelcomeBackMsg"];
+                tts.Speak(welcomeMsg, QueueMode.Flush, null, needUserAnswerId);
+            }
+        }
+
+        private void TtsStoppedSpeaking(object sender, UtteranceIdArgs e)
+        {
+            if (e.UtteranceId == needUserAnswerId)
+            {
+                voiceRecognizer.RecognizePhrase(this);
```

Comment:
So basically, if after a message we need to ask the user something, we do just that.